### PR TITLE
Add module `kleborate`

### PR DESCRIPTION
**Pull request by abhi18av**
## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by abhi18av** [comment]
@rpetit3 , regarding this one I am running into a bit of an issue writing tests. 

It seems that the path expected by `makeblastdb` isn't writeable. I've narrowed it down to the use of https://github.com/katholt/Kleborate/blob/eae8e794a77bbdf24a7c111ad8322d209daab742/kleborate/__main__.py#L366 

Also, I noticed that you've [recently patched](https://github.com/bioconda/bioconda-recipes/blob/b49874d4138e9997039a07b80e732d90a2c06f6d/recipes/kleborate/fix_get_kaptive_path.patch#L7) the bioconda package for `kleborate`, do you reckon that we should add a new parameter for allowing this `data` to be in a different position?  I'm not sure if it's possible to alter permissions using the `build.sh` on the bioconda recipe.



  

```

[c4/c242ef] process > test_kleborate:KLEBORATE (test) [100%] 1 of 1, failed: 1 ✘
Error executing process > 'test_kleborate:KLEBORATE (test)'

Caused by:
  Process `test_kleborate:KLEBORATE (test)` terminated with an error exit status (1)

Command executed:

  kleborate \
       \
      --outfile test.results.txt \
      --assemblies  *.fasta
  
  echo $(kleborate -v 2>&1) | sed 's/kleborate //;' > kleborate.version.txt

Command exit status:
  1

Command output:
  strain        species ST      virulence_score Yersiniabactin  YbST    Colibactin      CbST    Aerobactin      AbST    SalmochelinSmST     RmpADC  RmST    rmpA2   wzi     K_locus

Command error:
  Error: NCBI C++ Exception:
      T0 "/opt/conda/conda-bld/blast_1626166529662/work/blast/c++/src/objtools/blast/seqdb_writer/build_db.cpp", line 1053: Error: (CMultisourceException::eInvalid) BLASTDB::ncbi::CBuildDatabase::CreateDirectories() - You do not have write permissions on 'data'
  
  Traceback (most recent call last):
    File "/usr/local/bin/kleborate", line 10, in <module>
      sys.exit(main())
    File "/usr/local/lib/python3.8/site-packages/kleborate/__main__.py", line 61, in main
      results.update(get_all_virulence_results(data_folder, contigs, args))
    File "/usr/local/lib/python3.8/site-packages/kleborate/__main__.py", line 477, in get_all_virulence_results
      virulence_results = [get_ybt_mlst_results(data_folder, contigs, args),
    File "/usr/local/lib/python3.8/site-packages/kleborate/__main__.py", line 504, in get_ybt_mlst_results
      return get_virulence_cluster_results(data_folder, contigs, 'ybt_alleles.fasta',
    File "/usr/local/lib/python3.8/site-packages/kleborate/__main__.py", line 460, in get_virulence_cluster_results
      mlst_blast(seqs, database, 'yes', [contigs], min_cov=args.min_coverage,
    File "/usr/local/lib/python3.8/site-packages/kleborate/mlstBLAST.py", line 46, in mlst_blast
      hits = run_blastn(seqs, contigs, min_spurious_cov, min_spurious_ident)
    File "/usr/local/lib/python3.8/site-packages/kleborate/blastn.py", line 22, in run_blastn
      build_blast_database_if_needed(db)
    File "/usr/local/lib/python3.8/site-packages/kleborate/blastn.py", line 138, in build_blast_database_if_needed
      subprocess.check_call('makeblastdb -dbtype nucl -in ' + seqs, stdout=devnull,
    File "/usr/local/lib/python3.8/subprocess.py", line 364, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command 'makeblastdb -dbtype nucl -in /usr/local/lib/python3.8/site-packages/kleborate/data/ybt_alleles.fasta' returned non-zero exit status 255.

Work dir:
  /tmp/pytest_workflow_3la4lze1/kleborate/work/c4/c242efe06908cd9dfb240471a9d505

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`

```

**Comment by rpetit3** [comment]
@abhi18av is the test with Conda, Docker, or Singularity? I'll probably be able to play around with this tomorrow

**Comment by abhi18av** [comment]
The test was run using the `docker` profile 

```
PROFILE=docker pytest --tag kleborate --symlink --keep-workflow-wd
```

**Comment by rpetit3** [comment]
This one is going to be problematic because you cannot change the path of the database at run time (e.g. `--db /data/kleborate`).

Let me play around with the bioconda recipe to see if we can make it changed by an environmental variable (`${KLEBORATE_DB}`)

**Comment by rpetit3** [comment]
@abhi18av I submitted a PR to kleborate https://github.com/katholt/Kleborate/pull/59

I might be able to rebuild the bioconda recipe to include the blast indexes as a temporary patch

**Comment by rpetit3** [comment]
Also submitted recipe update to bioconda to build that blast databases https://github.com/bioconda/bioconda-recipes/pull/30582

**Comment by rpetit3** [comment]
Ok, I have a recipe update that pre-builds the BLAST databases. Hopefully this should allow you to continue

**Comment by abhi18av** [comment]
Thanks for the quick efforts on this @rpetit3! 

I think before giving finishing touches to this PR, it's best to await the merge for https://github.com/bioconda/bioconda-recipes/pull/30582 , right?

**Comment by rpetit3** [comment]
The new build was merged recently so it might take a little while for the containers to be built

**Comment by abhi18av** [comment]
@rpetit3 , thanks to the new build for `kleborate` I was able to get it moving fowards with the `docker` tests, however the `singularity` one is still failing. I don't have much experience with singularity (nor do I have it setup locally) - in case you have it handy, could you please give it a look?

If not, no worries, I'll give it a read and be up and running with singularity on a linux machine by tomorrow.

File: modules/kleborate/main.nf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    conda (params.enable_conda ? "bioconda::kleborate=2.1.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/kleborate:2.1.0--pyhdfd78af_1"
+    } else {
+        container "quay.io/biocontainers/kleborate:2.1.0--pyhdfd78af_1"
+    }
+
+    input:
+    tuple val(meta), path("*")
```
Comment (drpatelh) [inline modules/kleborate/main.nf:22]: What is this doing? Does it need to be `path(fastas)` instead?

File: modules/kleborate/main.nf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    kleborate \\
+        $options.args \\
+        --outfile ${prefix}.results.txt \\
+        --assemblies  *.fasta
```
Comment (drpatelh) [inline modules/kleborate/main.nf:35]:
```suggestion
        --assemblies *.fasta
```

File: tests/modules/kleborate/main.nf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+nextflow.enable.dsl = 2
+
+include { KLEBORATE } from '../../../modules/kleborate/main.nf' addParams( options: [:] )
+
+workflow test_kleborate {
+    
+    input = [ [ id:'test', single_end:false ], // meta map
+             [file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true),
+             file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true),
+             file(params.test_data['sarscov2']['illumina']['scaffolds_fasta'], checkIfExists: true)]]
```
Comment (drpatelh) [inline tests/modules/kleborate/main.nf:12]:
```suggestion
    input = [ 
        [ id:'test', single_end:false ], // meta map
        [ file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true),
          file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true),
          file(params.test_data['sarscov2']['illumina']['scaffolds_fasta'], checkIfExists: true)
        ]
    ]
```

File: modules/kleborate/main.nf (hunk 1 above)
Comment (drpatelh) [inline modules/kleborate/main.nf:22]:
```suggestion
    tuple val(meta), path(fastas)
```

**Comment by drpatelh** [comment]
Thanks @abhi18av @rpetit3 ! Will merge this in and hopefully the Singularity container will be pushed soon too. If not, we can raise an issue. They are generally very good.

**Review by drpatelh — APPROVED** [review]

**Comment by rpetit3** [comment]
Thank you @drpatelh ! many more in bound, looking forward to taking full advantage of `nf-core/modules`

**Comment by drpatelh** [comment]
Awesome! Thank you for reviewing too 😎 You know the ropes now.

**Pull request merged**